_drafts/practical_zephyr_03_dt_semantics.md

```diff
@@ -0,0 +1,1970 @@
+/ {
+  soc {
+    gpio0: gpio@50000000 {
+      compatible = "nordic,nrf-gpio";
+      #gpio-cells = <2>;
+      port = <0>;
+    };
+
+    gpio1: gpio@50000300 {
+      compatible = "nordic,nrf-gpio";
+      #gpio-cells = <2>;
+      port = <1>;
+    };
+  };
+};
+```
+
+Each instance handles _all_ pins in the corresponding port. However, some other Devicetree node, e.g., an LED, needs a way to control and configure a single pin of a given port. We can't do this using a normal _phandle_, e.g., `<&gpio0>`, we need a mechanisim to pass some _parameters_ or _metadata_ along with the _phandle_. `phandle-array`s implement this exact use case, allowing to pass a predefined number of cells with the _phandle_, e.g., as we can see in the nRF52840's development kit DTS file for `/leds/led_0`:
+
+`zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts`
+```dts
+/ {
+  leds {
+    led0: led_0 {
+      gpios = <&gpio0 13 GPIO_ACTIVE_LOW>;
+    };
+  };
+};
+```
+
+What we haven't seen yet, is how we can know that the first cell _13_ is the _pin_ within `gpio0`, and the second cell _GPIO_ACTIVE_LOW_ is a configuration _flag_. This is part of the _semantics_ defined by the _binding_ of the _gpio_ nodes that are compatible with `nordic,nrf-gpio`:
+
+`zephyr/dts/bindings/gpio/nordic,nrf-gpio.yaml`
+```yaml
+description: NRF5 GPIO node
+compatible: "nordic,nrf-gpio"
+# --snip--
+
+gpio-cells:
+  - pin
+  - flags
+```
+
+Here we see that `gpio-cells` consists of two items `pin` and `flags`, which are matched exactly against the provided metadata in the given order. Thus, _13_ refers to the `pin` and `GPIO_ACTIVE_LOW` to its `flags`.
+
+You may have noticed that the `gpio` nodes do not conform to the naming convention that we've seen in the last article. This makes it a bit awkward to explain what needs to be provided for a `phandle-array` and the referenced nodes. Therefore, we'll build our own from scratch.
+
+> **Note:** Have a look at [Zephyr's documentation about specifier cells](https://docs.zephyrproject.org/latest/build/dts/bindings-syntax.html#specifier-cell-names-cells) in case you want to know how `gpios` opts out of the mentioned naming convention.
+
+##### A `phandle-array` from scratch
+
+Let's briefly review the rules for using `phandle-array`s that we've seen in the previous article. Notice that there's no need to understand them out of context, we'll see all of them in action in our example:
+
+- By convention, a `phandle-array` property is plural and its name must thus end in _s_.
+- The value of a `phandle-array` property is an array of phandles, but each phandle is followed by the pre-defined number of cells for each referenced node.
+- The number of cells that can follow a node's reference is specified by the node's _specifier cells_ property `#<prefix>-cells`.
+- _Specifier cells_ have a defined naming convention: The name is formed by removing the plural '_s_' and attaching '_-cells_' to the name of the `phandle-array` property.
```

Comment:
I might combine these or switch the order, I got confused by seeing a prefix but that's not explicitly mentioned in either of the bullet points
```suggestion
- The number of cells that can follow a node's reference is specified by the node's _specifier cells_ property
- The _specifier cells_ property has a defined naming convention: The name is formed by removing the plural '_s_' and attaching '_-cells_' to the name of the `phandle-array` property: `#<name-without-s>-cells`
```